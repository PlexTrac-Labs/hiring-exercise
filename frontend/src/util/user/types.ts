
export type CreateUserRequest = {
    username: string;
    password: string;
    firstName: string;
    lastName: string;
    email: string;
    birthYear: number;
    favColor: string;
}

export type UpdateUserRequest = {
    username: string;
    firstName?: string;
    lastName?: string;
    email?: string;
    birthYear?: number;
    favColor?: string;
};

export type User = {
    _id: string;
    username: string;
    password: string;
    firstName: string;
    lastName: string;
    email: string;
    birthYear: number;
    favColor: string;
    admin: boolean;
}

export type AuthResponse = {
    auth_token: string;
    user: User;
}

export type ResetPasswordRequest = {
    oldPassword: string;
    newPassword: string;
}
