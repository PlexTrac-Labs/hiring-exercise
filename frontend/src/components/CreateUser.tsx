import { ListItem, ListItemButton, ListItemIcon } from "@mui/material";
import React from "react";
import { LIST_MIN_WIDTH } from "../helpful/constants";
import AddCircleIcon from "@mui/icons-material/AddCircle";

export const CreateUser = () => {
  const style = {
    minWidth: LIST_MIN_WIDTH,
  };

  return (
    <ListItem sx={style}>
      <ListItemButton>
        <ListItemIcon>
          <AddCircleIcon color="success" />
        </ListItemIcon>
        Create User
      </ListItemButton>
    </ListItem>
  );
};
