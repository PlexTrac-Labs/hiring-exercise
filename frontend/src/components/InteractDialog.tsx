import { Dialog, DialogContent, DialogTitle } from "@mui/material";
import React from "react";
import { colors } from "../helpful/style";

export type InteractDialogProps = {
  open: boolean;
  title: string;
  handleClose: () => void;
};

export const InteractDialog: React.FC<InteractDialogProps> = (props) => {
  const { children, open, title, handleClose } = { ...props };

  const TitleStyle = {
    backgroundColor: colors.darkGreen,
    color: "white",
  };

  return (
    <Dialog open={open} onClose={handleClose}>
      <DialogTitle sx={TitleStyle}>{title}</DialogTitle>
      <DialogContent>{children}</DialogContent>
    </Dialog>
  );
};
