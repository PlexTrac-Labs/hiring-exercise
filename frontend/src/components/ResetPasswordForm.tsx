import { Alert, Button, FormControl, Grid, TextField } from "@mui/material";
import React from "react";

export type ResetFormProps = {
  error: boolean;
  errorMessage: string;
  handleSubmit: (event: any) => void;
};

const defaultProps: Partial<ResetFormProps> = {};

export const ResetPasswordForm = (props: ResetFormProps) => {
  const { error, errorMessage, handleSubmit } = {
    ...defaultProps,
    ...props,
  };
  return (
    <form onSubmit={handleSubmit}>
      <Grid
        container
        alignItems="center"
        justifyContent="center"
        rowSpacing={2}
        marginTop={3}
      >
        {error && (
          <Grid item xs={12}>
            <Alert severity="error">{errorMessage}</Alert>
          </Grid>
        )}
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              type="password"
              id="old"
              variant="outlined"
              label="Current Password"
            />
          </FormControl>
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              type="password"
              id="new"
              variant="outlined"
              label="New Password"
            />
          </FormControl>
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              type="password"
              id="confirm"
              variant="outlined"
              label="Confirm New Password"
            />
          </FormControl>
        </Grid>
        <Grid item xs={12}>
          <Button color="primary" variant="contained" type="submit">
            Change Password
          </Button>
        </Grid>
      </Grid>
    </form>
  );
};
