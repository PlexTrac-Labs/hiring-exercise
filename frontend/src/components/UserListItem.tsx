import DeleteIcon from "@mui/icons-material/Delete";
import {
  IconButton,
  ListItem,
  ListItemButton,
  ListItemText,
  Tooltip,
} from "@mui/material";
import React from "react";
import { LIST_MIN_WIDTH } from "../helpful/constants";
import { User } from "../util/user/types";

export type UserListItemProps = {
  user: User;
  handleDelete: () => void;
  handleClick: (user: User) => void;
};

export const UserListItem = (props: UserListItemProps) => {
  const { user, handleDelete, handleClick } = { ...props };

  return (
    <ListItem
      sx={{ minWidth: LIST_MIN_WIDTH }}
      secondaryAction={
        <Tooltip title="Delete User" placement="right" enterDelay={1000}>
          <IconButton edge="end" onClick={handleDelete}>
            <DeleteIcon />
          </IconButton>
        </Tooltip>
      }
    >
      <ListItemButton onClick={() => handleClick(user)}>
        <ListItemText>{user.username}</ListItemText>
      </ListItemButton>
    </ListItem>
  );
};
