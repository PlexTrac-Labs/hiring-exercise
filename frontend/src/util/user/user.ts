import axios from "axios";
import { CreateUserRequest, UpdateUserRequest } from "./types"


// POST
export const createUser = (payload: CreateUserRequest) => {
    console.log('kdp createUser');
}

export const authenticate = (username: string, password: string) => { //kdp revisit these params
    console.log('kdp authenticate');
    return axios.post(`http://localhost:5000/authenticate`, { username, password})
        .then(res => res.data);
}

// PUT
export const updateUser = (payload: UpdateUserRequest) => {
    console.log('kdp update user')
}

//DELETE
export const deleteUser = (id: string) => {
    console.log('kdp deleteUser')
}

// GET

export const getUser = (id: string) => {
    return axios.get(`http://localhost:5000/user/${id}`)
        .then(res => res.data)
}

export const getAllUsers = () => {
    return axios.get(`http://localhost:5000/user`)
    .then(res => res.data)
}