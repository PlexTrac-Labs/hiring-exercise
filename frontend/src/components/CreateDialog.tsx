import React, { useState } from "react";
import { CreateUserRequest, User } from "../util/user/types";
import { createUser, getAllUsers } from "../util/user/user";
import { InteractDialog } from "./InteractDialog";
import { UserForm } from "./UserForm";

export type CreateDialogProps = {
  open: boolean;
  setUsers: (users: User[]) => void;
  onClose: () => void;
};

export const CreateDialog = (props: CreateDialogProps) => {
  const { open, setUsers, onClose } = { ...props };
  const [error, setError] = useState(false);
  const errorMessage = "Problem creating user. Is this email already in use?";

  const handleClose = () => {
    setError(false);
    onClose();
  };

  const handleCreate = async (event: any) => {
    event.preventDefault();
    const payload: CreateUserRequest = {
      username: event.target.username.value,
      password: event.target.password.value,
      firstName: event.target.firstName.value,
      lastName: event.target.lastName.value,
      email: event.target.email.value,
      birthYear: parseInt(event.target.birthYear.value),
      favColor: event.target.favColor.value,
    };
    try {
      await createUser(payload);
      let users = await getAllUsers();
      setUsers([...users]);
      handleClose();
    } catch (error: any) {
      setError(true);
    }
  };

  return (
    <InteractDialog open={open} handleClose={handleClose} title="Create User">
      <UserForm
        showPassword
        submitText="Create"
        error={error}
        errorMessage={errorMessage}
        handleSubmit={handleCreate}
        handleClose={handleClose}
      />
    </InteractDialog>
  );
};
