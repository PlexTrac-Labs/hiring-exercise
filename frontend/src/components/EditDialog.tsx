import React, { useState } from "react";
import { UpdateUserRequest, User } from "../util/user/types";
import { getAllUsers, updateUser } from "../util/user/user";
import { InteractDialog } from "./InteractDialog";
import { UserForm } from "./UserForm";

export type EditDialogProps = {
  open: boolean;
  user: User;
  setUsers: (users: User[]) => void;
  onClose: () => void;
};

export const EditDialog = (props: EditDialogProps) => {
  const { open, user, setUsers, onClose } = { ...props };
  const [error, setError] = useState(false);
  const errorMessage = "Problem updating user. Is this email already in use?";

  const handleClose = () => {
    setError(false);
    onClose();
  };

  const handleUpdate = async (event: any) => {
    event.preventDefault();
    const payload: UpdateUserRequest = {
      username: event.target.username.value,
      firstName: event.target.firstName.value,
      lastName: event.target.lastName.value,
      email: event.target.email.value,
      birthYear: parseInt(event.target.birthYear.value),
      favColor: event.target.favColor.value,
    };
    try {
      await updateUser(user._id, payload);
      let users = await getAllUsers();
      setUsers([...users]);
      handleClose();
    } catch (error: any) {
      setError(true);
    }
  };

  return (
    <InteractDialog open={open} handleClose={handleClose} title="Edit User">
      <UserForm
        user={user}
        submitText="Save"
        error={error}
        errorMessage={errorMessage}
        handleSubmit={handleUpdate}
        handleClose={handleClose}
      />
    </InteractDialog>
  );
};
