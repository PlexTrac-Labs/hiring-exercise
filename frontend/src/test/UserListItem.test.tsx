import { cleanup, render } from "@testing-library/react";
import React from "react";
import { UserListItem, UserListItemProps } from "../components/UserListItem";
import { User } from "../util/user/types";
import { EMPTY_USER } from "./testConstants";

const props: UserListItemProps = {
  user: EMPTY_USER,
  handleClick: (user: User) => {},
  handleDelete: () => {},
};

afterEach(cleanup);

it("renders", async () => {
  const { asFragment } = render(<UserListItem {...props} />);
  expect(asFragment()).toMatchSnapshot();
});
