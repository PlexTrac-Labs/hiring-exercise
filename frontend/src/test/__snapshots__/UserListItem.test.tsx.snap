// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <li
    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-f73flm-MuiListItem-root"
  >
    <div
      class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        />
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
    <div
      class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
    >
      <button
        aria-label="Delete User"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element="true"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </li>
</DocumentFragment>
`;
